### ISSUE TITLE
Moya on Swift 2.3 & 3.0.

### ISSUE THREAD

**Issue by aaroncrespo** [Moya/Moya#513]
Took a crack at checking out swift 2.3 support this morning... 
### The good news:
- Apple's migrator tool is great, after running on Moya, the demo target and tests (iOS,tvOS)  built and worked/passed
- The changes don't look too bad from Moya's end
- Foundation Value types and the new Swift API Guidelines don't come into play until Swift 3 which greatly limits changes.
- swift 2 and 2.3 can live side by side with `#if swift(>=2.3) ...` 
### The meh news:
- Nimble under macOS had some issues I've seen reported as Xcode/Foundation related -- affects macOS test target
- xcodebuild doesn't pick up the swift legacy flag which triggers a compile against a 2.0 toolchain Xcode _should_ in the future hopefully not do this. Until then `rake build` and `rake test` don't work without setting the TOOLCHAINS Environment flag. This is probably a bug in xcodebuild.
- Relatedly Setting of the legacy swift flag is not performed by the migrator maybe because they want 2.0/2.3/3.0 to try and live side by side but that will cause weirdness with dependent projects.

So tracking what happened when I got it all building:
- _Moya Changes:_ 
- [ ] A couple URL API's changed nullability. Should check documentation and expectations and see if implicitly unwrapped is safe. Most are similar to:

``` diff
 public func url(route: TargetType) -> String {
-    return route.baseURL.URLByAppendingPathComponent(route.path).absoluteString
+    return route.baseURL.URLByAppendingPathComponent(route.path)!.absoluteString!
 }
```
- [x] Decision on setting Swift Legacy flags. 2.0 and 2.3 coexisting is great, but value types + API guideline changes are not a small change. (probably breaking or a lot of fixme writing)
- _Dependencies_
- [ ] Alamofire:  https://github.com/Alamofire/Alamofire/pull/1313
- [ ] Result: GTG
- [x] Quick/Nimble: https://github.com/Quick/Quick/issues/55, https://github.com/Quick/Nimble/pull/302
- [ ] RxSwift: ??
- [ ] ReactiveCocoa: ?? some small attribute updates -- Don't know if they have an issue yet.

P.S. It's been a while since I ran Moya in dev mode and it took me a little while to get oriented, do you guys think a quick _Getting Started With Developing Moya_ section somewhere would be useful? maybe #270. I don't really use rake for build scripting so it took a bit to get that setup.

**Comment by brentleyjones** [comment]
> RxSwift: ??

With ReactiveX/RxSwift@9b1583f2e365ef78d250be278ca5838076b58095 RxSwift supports Swift 2.3. It doesn't have the correct build flag.

**Comment by ashfurrow** [comment]
This is great work and a nice comprehensive write-up, thanks @aaroncrespo!

**Comment by aaroncrespo** [comment]
Another thing to think about is if it makes sense to skip 2.3 and go straight to 3.0 or how to support 3 versions. The `#if swift()` tests don't work so well when swift 3.0 is in the mix.

**Comment by lowell** [comment]
Sorry about the accidental closure. Misunderstood what 'Close' meant in a third-party GitHub client. Was just trying to close the issue view. 😳

**Comment by ashfurrow** [comment]
No problem, it happens!

**Comment by brentleyjones** [comment]
I highly recommend a Swift 2.3 build, even as a version bump, before going to Swift 3. Alamofire had breaking changes so without this people won't be able to use pre-Swift 3 on Xcode 8. I know I'll be using Xcode 8 and Swift 2.3 for a while after the release.

**Comment by sunshinejr** [comment]
I've changed the title due to #522 discussion.

**Comment by orta** [comment]
Aye, think it's worth shipping a 2.3 dupe of current Moya, one the Rx deps have them, then moving to 3.0

**Comment by ashfurrow** [comment]
That makes sense – should we make 2.3 version off of `master` or the most recent commit? (master has breaking changes.)

**Comment by orta** [comment]
Maybe the last stable release?

**Comment by aaroncrespo** [comment]
Looks like Alamofire is about to cut a `3.5.0` release which will have both swift 2.2 and swift 2.3 support. Then Moya.spec can point to that.

https://github.com/Alamofire/Alamofire/milestone/27

I think this is the  dependency to push a release supporting 2.3

**Comment by tdouglas1313** [comment]
Is there any update on swift 3.0 support now that Xcode 8 GM is out

**Comment by tommy60703** [comment]
Need swift 2.3 and swift 3.0 support here...

**Comment by sunshinejr** [comment]
There is a branch for swift 3.0 already, but needs update. I'm downloading GM right now and will update the branch ASAP.

**Comment by BasThomas** [comment]
@sunshinejr It seems Result already is Swift 3. :)

**Comment by AndrewSB** [comment]
This looks like it's been solved with the 8.0.0 beta releases, I'm going to close it out. Please feel free to open this if there are any loose ends!
